### PR TITLE
Added dist functionality to torch reduction operations

### LINKED ISSUE #4046: dist

### PR DESCRIPTION
**Pull request by sayantan1410**
Fixes #4046 
I am getting an attribute error while testing the function because the backend that is being used is numpy and I want it to be torch. 
Let me know if there's any more error.

### CODE REVIEW

File: ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+):
+    input_x_dtype, x, dim_y = dtype_and_x
+    input_y_dtype, y, dim_y = dtype_and_y
+    helpers.test_frontend_function(
+        input_dtypes=[input_x_dtype, input_y_dtype],
+        as_variable_flags=as_variable,
+        with_out=False,
+        num_positional_args=num_positional_args,
+        native_array_flags=native_array,
+        fw="torch",
```
Comment (abdrahmandiab) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py:37]: you should set fw here to be equal to fw. This test is for using torch as a frontend, however it should test this frontend functionality with all backends! 😅

File: ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+@handle_cmd_line_args
+@given(
+    dtype_and_x=helpers.statistical_dtype_values(function="dist"),
+    dtype_and_y=helpers.statistical_dtype_values(function="dist"),
+    as_variable=helpers.array_bools(num_arrays=2),
+    num_positional_args=helpers.num_positional_args(
+        fn_name="ivy.functional.frontends.torch.dist"
+    ),
+    native_array=helpers.array_bools(num_arrays=2),
+    p=st.integers(),
```
Comment (abdrahmandiab) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py:18]: please use the hypothesis strategy of 
st.sampled_from( st.integers(), None) , this will enable hypothesis to generate None values for p since it is optional.

File: ivy/functional/frontends/torch/reduction_ops.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import ivy
+
+
+def dist(input, other, p=2):
+    return ivy.dist(input, other, p)
```
[the remark this answers is not in the document]
Comment (abdrahmandiab) [reply ivy/functional/frontends/torch/reduction_ops.py:5]: An example code to use here could be smth like
ivy.pow( ivy.sum(ivy.pow(ivy.abs(x-y),p)),1/p)

This function call should return the p-norm value given x and y

**Review by abdrahmandiab — CHANGES_REQUESTED** [review]
Some minor changes required before we can fix the test :)

File: ivy/functional/frontends/torch/reduction_ops.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sayantan1410) [reply ivy/functional/frontends/torch/reduction_ops.py:5]: Thank you so much, can i not use ivy.norm ?

File: ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py (hunk 1 above)
Comment (sayantan1410) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py:37]: Oh I see !!

File: ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py (hunk 2 above)
Comment (sayantan1410) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py:18]: Okay !

File: ivy/functional/frontends/torch/reduction_ops.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abdrahmandiab) [reply ivy/functional/frontends/torch/reduction_ops.py:5]: aah... I am not quite sure that this function exists either unfortunately 😅 But, sure if you find another function that implements the behavior of dist correctly, then that is all that matters.

**Comment by sayantan1410** [comment]
@abdrahmandiab  Hey, I have implemented the dist function and looks like it passes most of the test cases, right now when one array cannot be broadcasted to another that is when the tests are failing.

**Comment by sayantan1410** [comment]
@abdrahmandiab  Hey,  Right now all the tests are passing when I am running them locally, I was only getting an warning saying: 
`RuntimeWarning: overflow encountered in power` , which I have silenced using filter warning, other than can you please review it once and let me what to fix.

**Comment by abdrahmandiab** [comment]
hi @sayantan1410 , I actually just realized the function you are looking for is likely [ivy.vector_norm](https://github.com/unifyai/ivy/blob/87edb17af8a379dd9286581fe04652e936a69c7c/ivy/functional/ivy/linear_algebra.py#L1553). 

Could you please use this function instead in your frontend implementation? 😅

Thanks, and apologies for the inconvenience! Good job so far though!

**Comment by sayantan1410** [comment]
Hey, @abdrahmandiab I resolved the conflicts and all the frontend tests are passing but "check for failures" is failing and I cannot figure out why. Can you help me on this as well ?

**Comment by abdrahmandiab** [comment]
> Hey, @abdrahmandiab I resolved the conflicts and all the frontend tests are passing but "check for failures" is failing and I cannot figure out why. Can you help me on this as well ?

Hi @sayantan1410,
I'm sorry if the CI is a bit confusing at the moment, we're experimenting with some things! 😅 
To check for why the tests are failing please check the "Run Frontend Tests" job's output rather than "Check for Failures"

There are errors by functions other than dist, but you can search for errors thrown by torch dist by searching the logs similar to the image below. 😃 
image

Thanks!

File: ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py [hunk 4]
```diff
@@ -1,12 +1,64 @@
...
+            pass
+        else:
+            return False
+    return True
+
+
+@pytest.mark.filterwarnings("ignore: dist")
+@handle_cmd_line_args
+@given(
+    dtype_and_x=statistical_dtype_values(function="dist"),
```
Comment (abdrahmandiab) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py:26]: Please use the helper function dtype_and_values() to generate the data :)

In this case you should use the following settings
```
helpers.dtype_and_values(
available_dtypes = helpers.dtypes("numeric")
num_arrays = 2,
shared_dtype=True,
allow_inf=False,
)
```
This should generate 2 arrays of similar shape that you can use instead of x and y. :)

This code ^^ should generally be correct but please feel free to modify it to suit the function.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py [hunk 5]
```diff
@@ -1,12 +1,64 @@
...
+def is_broadcastable(shp1, shp2):
+    for a, b in zip(shp1[::-1], shp2[::-1]):
+        if a == 1 or b == 1 or a == b:
+            pass
+        else:
+            return False
+    return True
+
+
+@pytest.mark.filterwarnings("ignore: dist")
```
Comment (abdrahmandiab) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py:23]: why was this added?

**Review by abdrahmandiab — CHANGES_REQUESTED** [review]
tiny changes to be made here. Otherwise everything looks good! 😄 

Thanks!

File: ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py (hunk 5 above)
Comment (sayantan1410) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py]: Sorry ! this was by mistake, initially I was getting some warnings so I added this line to silence them.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py (hunk 4 above)
Comment (sayantan1410) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_reduction_ops.py]: Yeah, It worked well, I only had to change `helpers.dtypes`  to `helpers.get_dtypes` as I got an error saying `module 'ivy_tests.test_ivy.helpers' has no attribute 'dtypes'`.

**Comment by sayantan1410** [comment]
Hello @abdrahmandiab , I fixed the things that you suggested. The lint tests are failing for some inconsistencies in the miscellaneous_ops.py file.

**Comment by abdrahmandiab** [comment]
Looks good to me @sayantan1410,
Thanks for your hard work! 😄

**Comment by sayantan1410** [comment]
Thank you for all the help, it was great working alongside !
